### ISSUE TITLE
Proxy options with Xchange

### ISSUE THREAD

**Issue by archenroot** [timmolter/XChange#2225]
Hi guys, in situation like being on production system without direct connectivity to internet, what are proxy options with xchange?

Also I heard that exchanges might block connections detected coming from proxies, is there any way to clean up proxies from being detected?

Thanks for any hints.

Ladislav

**Comment by daniloguimaraes** [comment]
@archenroot , XChange has a "built-in" option for proxy at [`ExchangeSpecification`](https://github.com/timmolter/XChange/blob/develop/xchange-core/src/main/java/org/knowm/xchange/ExchangeSpecification.java) . Fields `String proxyHost` and `Integer proxyPort`.

These properties are used internally by [rescu library](https://github.com/mmazi/rescu) at [`BaseExchangeService`](https://github.com/timmolter/XChange/blob/develop/xchange-core/src/main/java/org/knowm/xchange/service/BaseExchangeService.java) class

In my humble opinion, these 2 properties aren't enough, because sometimes we face an authenticated proxy. So, I would add `String proxyUsername` and `String proxyPassword` properties at [`ExchangeSpecification`](https://github.com/timmolter/XChange/blob/develop/xchange-core/src/main/java/org/knowm/xchange/ExchangeSpecification.java)

Also, you could test the [Java Networking Properties](https://docs.oracle.com/javase/7/docs/api/java/net/doc-files/net-properties.html) at your application and hopely, it will works.

**Comment by archenroot** [comment]
@daniloguimaraes Thank you very much for confirmation and hint. I will test it carefully. I was already under impression of using standard:
```
System.setProperty("http.proxyHost", "");
System.setProperty("http.proxyPort", "");
```
or using OS level proxy settin (can be set per user), but both actually do JVM wide setting.

The way you describe is even more robust, because you can create 2 exchange instances, each with its own proxy within the same JVM, is my assumption correct?

Best regards,

Ladislav

**Comment by daniloguimaraes** [comment]
Yeah, @archenroot . When you use `System.setProperty`, will be applied to entire JVM.

Now, using specific exchange specification you can mix proxies (even no proxy).

Good luck.

**Comment by archenroot** [comment]
@daniloguimaraes Thx budy, I actually implemented new method in rescu code:
https://github.com/mmazi/rescu/issues/104

But I am not sure if that is enough or not.

I also extended in Xchange code ExchangeSpecification by adding proxyUser and proxyHost, but where is the exact place to inject the header? I am quite new to this codebase and although it is well documented, I am not sure simply :-)

**Comment by daniloguimaraes** [comment]
@archenroot , I suppose a simple example could be like this: 

```java
Exchange yourFavoriteExchange = ExchangeFactory.INSTANCE.createExchange(YourFavoriteExchange.class.getName());

ExchangeSpecification exchangeSpec = yourFavoriteExchange.getDefaultExchangeSpecification();

exchangeSpec.setProxyHost("localhost");
exchangeSpec.setProxyPort(8080);

yourFavoriteExchange.applySpecification(exchangeSpec);

// Now your 'yourFavoriteExchange' instance is ready to connect through proxy.
```

**Comment by archenroot** [comment]
@daniloguimaraes Well, this is easy one, which works well already, but the extension with ProxyAuthorization HTTP header, where to put it?

**Comment by daniloguimaraes** [comment]
There is method in `BaseExchangeService` called `getClientConfig()` ([check here](https://github.com/timmolter/XChange/blob/e1ae0c4b45af867b1f1ed3337037867150e23046/xchange-core/src/main/java/org/knowm/xchange/service/BaseExchangeService.java#L54))

This method already set up proxy host and port, if needed.
You could open a PR to add proxy username and password as well.

**Comment by archenroot** [comment]
@daniloguimaraes - I will open PR definitely I work on fork already, I think the change will need to happen in rescu lib. But still investigating

**Comment by FedrBodr** [comment]
@archenroot how about the result of your research? Now I am got the same issue) And i start my own  investigating

**Comment by aacdevelop** [comment]
Hi, There's a way to set also username and password in the proxy settings?

**Comment by bbottema** [comment]
Any update here? I also want to connect via authenticated proxy.

**Comment by mapleafly** [comment]
My Xchange is 5.0.0-SNAPSHOT. My code is ok.
        ExchangeSpecification exSpec = new BinanceExchange().getDefaultExchangeSpecification();
        exSpec.setProxyHost("127.0.0.1");//edit youself
        exSpec.setProxyPort(50772);////edit youself
        Exchange exchange = ExchangeFactory.INSTANCE.createExchange(exSpec);

**Comment by elkceo** [comment]
> @daniloguimaraes - I will open PR definitely I work on fork already, I think the change will need to happen in rescu lib. But still investigating

Hi. Did you get solution for entering proxy credentials to exchange specification?

**Comment by wodepig** [comment]
I also encountered the same problem, how to set up a proxy with' username' and' password'
